feat(stark-build): implement integration with Angular CLI via @angular-builders

christophercr opened the pull request:
ISSUES CLOSED: #883, #146, #114 

## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/NationalBankBelgium/stark/blob/master/CONTRIBUTING.md#-commit-message-guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [X] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[X] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[X] Build related changes
[X] CI related changes
[X] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->
Stark based apps are not compatible with Angular CLI due to the customizations done in Stark-Build.

Issue Number: #883, #146, #114 

## What is the new behavior?
Stark-Build now overrides and expands the default build configuration from Angular CLI to leverage the functionalities needed for the different Stark packages.

## Does this PR introduce a breaking change?
```
[X] Yes
[ ] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

SuperITMan commented on starter/package.json:
```diff
@@ -26,13 +26,15 @@
   },
   "scripts": {
-    "build:aot:prod": "npm run clean:dist && npm run clean:aot && cross-env BUILD_AOT=1 npm run webpack -- --config node_modules/@nationalbankbelgium/stark-build/config/webpack.prod.js  --progress --profile --bail",
+    "build:aot:prod": "npm run clean:dist && npm run clean:aot && npm run ng -- build --aot=true --prod=true --buildOptimizer=true",
```
Same here about the arguments `--aot` and `--buildOptimizer`

SuperITMan commented on showcase/src/index.html:
```diff
@@ -34,11 +34,11 @@
-		<% for (let i = 0; i < htmlWebpackPlugin.options.dllFiles.css.length; ++i) { %>
-		<link href="<%= htmlWebpackPlugin.options.dllFiles.css[i] %>" rel="stylesheet" />
-		<% } %> <% } %>
+		<!--<% if (htmlWebpackPlugin.options.dllFiles) { %>-->
```
Do we plan to have support for dll files later ?

SuperITMan commented on packages/stark-build/config/build-utils.js:
```diff
@@ -20,11 +20,11 @@ const ANGULAR_APP_CONFIG = {
+	IS_DEV_SERVER: helpers.hasProcessFlag("serve"),
+	HMR: helpers.hasProcessFlag("hmr"),
+	AOT: helpers.hasProcessFlag("aot"),
+	E2E: helpers.hasProcessFlag("e2e"),
```
I don't see anywhere in our current `package.json` the flag `--e2e`. Can you confirm it is correct ? 😊

christophercr replied:
Indeed, the flag is not used... I kept it just in case... we should find a way to know whether the build is for e2e... but I think we can do that later

SuperITMan commented on packages/stark-build/config/webpack-partial.dev.js:
```diff
@@ -0,0 +1,118 @@
+	PORT: process.env.PORT || 4200,
+	ENV: isProd ? "production" : "development",
+	environment: helpers.hasProcessFlag("hmr") ? "hmr" : "dev",
+	WATCH: !helpers.hasProcessFlag("watch=false"),
```
Why keeping this logic in `build-utils.js` and repeating it here ?

christophercr replied:
True, I'll remove it from here

SuperITMan commented on packages/stark-build/config/webpack-partial.prod.js:
```diff
@@ -0,0 +1,36 @@
+	PORT: process.env.PORT || 4200,
+	ENV: isProd ? "production" : "development",
+	environment: helpers.hasProcessFlag("e2e") ? "e2e.prod" : "prod",
+	IS_DEV_SERVER: false
```
Why keeping this logic in `build-utils.js` and repeating it here ?

christophercr commented on showcase/angular.json:
```diff
@@ -88,7 +115,7 @@
-          "builder": "@angular-devkit/build-angular:dev-server",
+          "builder": "@angular-builders/dev-server:generic",
           "options": {
             "browserTarget": "showcase:build"
```
Mmm I was hesitating about that... and now that I think of it, I don't see why why should use a different port from the default one that NG CLI uses, don't you think?

On the other hand, I do find it useful to open automatically the browser, so I'll that option

christophercr commented on showcase/config/webpack-custom-config.dev.json:
```diff
@@ -1,4 +1,4 @@
 {
-  "cspConnectSrc": "http://localhost:5000 http://localhost:5001 http://localhost:5002 http://localhost:4000 https://nationalbankbelgium.github.io",
+  "cspConnectSrc": "http://localhost:5000 http://localhost:5001 http://localhost:5002 http://localhost:4000 https://nationalbankbelgium.github.io ws://localhost:4200",
```
In fact, that was my bad... I think we should not modify the `HOST` and `PORT` via environment variables anymore because NG CLI already allows this via an option

SuperITMan replied:
If we can get the default port and default host or the values defined in **angular.json** (if defined), I think it would be better to keep the previous scenario in `cspDirectives` in webpack-partial.dev.js --> `"connect-src 'self' ws://" + METADATA.HOST + ":" + METADATA.PORT + " " " + webpackCustomConfig["cspConnectSrc"]`

christophercr replied:
True, I'm fixing this

christophercr commented on packages/stark-build/config/webpack-partial.prod.js:
```diff
@@ -0,0 +1,36 @@
+
+const METADATA = Object.assign({}, buildUtils.DEFAULT_METADATA, {
+	HOST: process.env.HOST || "localhost",
+	PORT: process.env.PORT || 4200,
```
I don't think we need this anymore... we should remove it and simply use the NG CLI to define the Host and Port. There is no reason to keep this ourselves

christophercr commented on showcase/package.json:
```diff
@@ -26,16 +26,15 @@
   "scripts": {
-    "build:aot:prod": "npm run clean:dist && npm run clean:aot && cross-env BUILD_AOT=1 npm run webpack -- --config node_modules/@nationalbankbelgium/stark-build/config/webpack.prod.js  --progress --profile --bail",
-    "build:aot:prod:ci-test-env": "npm run clean:dist && npm run clean:aot && cross-env BUILD_AOT=1 npm run webpack -- --config node_modules/@nationalbankbelgium/stark-build/config/webpack.prod.js  --progress --profile --bail --ci-test-env",
+    "build:aot:prod": "npm run clean:dist && npm run clean:aot && npm run ng -- build --aot=true --prod=true --buildOptimizer=true",
```
Question... what is the preferred way to define this? Via the CLI? or via the `angular.json`?

christophercr replied:
Now that I think about it, we should get the value of this props from 2 places: the NG CLI and the `angular.json`... so it works always regardless of how you set such options.

I'll adapt the code in `build-utils`

christophercr commented on docs/stark-core/GETTING_STARTED.md:
```diff
@@ -0,0 +1,79 @@
+|
++---yourApp
+|   +---assets
+|	|	+---stark-core  // assets from Stark-Core should be available at this location
```
Nop, I'll fix it ;)

christophercr commented on docs/stark-ui/GETTING_STARTED.md:
```diff
@@ -76,3 +76,89 @@ For this reason, you must add these typings to the `typeRoots` in the `tsconfig.
+|
++---yourApp
+|   +---assets
+|	|	+---stark-ui  // assets from Stark-UI should be available at this location
```
Nop, I'll fix it ;)

christophercr commented on packages/stark-build/config/webpack-partial.common.js:
```diff
@@ -0,0 +1,297 @@
+	// env MONITOR variable set via cross-env
+	if (process.env.MONITOR) {
+		MONITOR = !!process.env.MONITOR;
+	}
```
True, it is cleaner... I'll fix it ;)

christophercr commented on packages/stark-build/config/webpack-partial.common.js:
```diff
@@ -0,0 +1,297 @@
+	// env BUNDLE_ANALYZER variable set via cross-env
+	if (process.env.BUNDLE_ANALYZER) {
+		BUNDLE_ANALYZER = !!process.env.BUNDLE_ANALYZER;
+	}
```
True, it is cleaner... I'll fix it ;)

cnomes commented:
@christophercr, I was wrong it is the update of `@mdi/material-icons`

christophercr commented on showcase/angular.json:
```diff
@@ -19,18 +28,28 @@
                 "glob": "**/*",
-                "input": "assets",
-                "output": "./assets"
+                "input": "node_modules/@nationalbankbelgium/stark-core/assets/",
```
Nop, not needed ;)

christophercr commented on showcase/angular.json:
```diff
@@ -19,18 +28,28 @@
                 "glob": "**/*",
-                "input": "assets-base",
-                "output": "./"
+                "input": "node_modules/@nationalbankbelgium/stark-ui/assets/",
```
Nop, not needed ;)

christophercr commented:
@SuperITMan @cnomes I've updated my PR with your remarks. I also fixed the warning about `(3726:7) start value has mixed support, consider using flex-start instead` .

Can you review this PR again?

christophercr commented on docs/stark-build/NG_CLI_BUILD_CUSTOMIZATIONS.md:
```diff
@@ -0,0 +1,302 @@
+            "serve": {
+                "builder": "@angular-builders/dev-server:generic",
+                "options": {
+                    "browserTarget": "your-app:build"
```
I don't think it is necessary to add it in this doc, the host and port are just app specific so in Stark we don't really care about it

christophercr commented:
@cnomes it seems that they noticed the version 3.5.94 from `@mdi/material-icons` was buggy so they un-released it :P

#1151 
https://travis-ci.org/NationalBankBelgium/stark/builds/498287079
https://www.npmjs.com/package/@mdi/angular-material

SuperITMan commented on showcase/angular.json:
```diff
@@ -88,9 +113,11 @@
+            "open": true
           },
           "configurations": {
             "hmr": {
```
To make hmr working, please set it like this (according to the official [doc](https://github.com/angular/angular-cli/wiki/stories-configure-hmr)):
```json
"hmr": {
  "hmr": true,
  "browserTarget": "showcase:build:hmr"
}
```

SuperITMan commented on starter/angular.json:
```diff
@@ -84,9 +109,11 @@
+            "open": true
           },
           "configurations": {
             "hmr": {
```
To make hmr working, please set it like this (according to the official [doc](https://github.com/angular/angular-cli/wiki/stories-configure-hmr)):
```json
"hmr": {
  "hmr": true,
  "browserTarget": "showcase:build:hmr"
}
```

SuperITMan commented:
Also, in the commit message, don't forget the mention to BREAKING CHANGE. Otherwise, it won't appear in our CHANGELOG :confused:

christophercr commented on packages/stark-build/config/webpack-partial.dev.js:
```diff
@@ -0,0 +1,121 @@
+const METADATA = Object.assign({}, buildUtils.DEFAULT_METADATA, {
+	HOST: buildUtils.ANGULAR_APP_CONFIG.config.architect.serve.options.host || helpers.getNgCliCommandOption("host") || "localhost", // by default is "localhost"
+	PORT: buildUtils.ANGULAR_APP_CONFIG.config.architect.serve.options.port || helpers.getNgCliCommandOption("port") || 4200, // by default is 4200
+	ENV: isProd ? "production" : "development",
```
Nop, they are different.... in DEV they are taken from `buildUtils.ANGULAR_APP_CONFIG.buildOptions` and in PROD from `buildUtils.ANGULAR_APP_CONFIG.buildConfigurations.production`

christophercr commented on packages/stark-build/config/webpack-partial.prod.js:
```diff
@@ -0,0 +1,38 @@
+	HOST: buildUtils.ANGULAR_APP_CONFIG.config.architect.serve.options.host || helpers.getNgCliCommandOption("host"),
+	PORT: buildUtils.ANGULAR_APP_CONFIG.config.architect.serve.options.port || helpers.getNgCliCommandOption("port"),
+	ENV: "production",
+	BASE_URL: buildUtils.ANGULAR_APP_CONFIG.buildConfigurations.production.baseHref,
```
Nop, they are different.... in DEV they are taken from `buildUtils.ANGULAR_APP_CONFIG.buildOptions` and in PROD from `buildUtils.ANGULAR_APP_CONFIG.buildConfigurations.production`

christophercr commented:
@cnomes @SuperItMan I've updated the PR with the latest remarks. Could you please have a look?

christophercr commented:
And.. I've updated the commit message to add the breaking changes 😉
